### DISCUSSION TITLE
PKCS#11 Key Generation with Safenet eToken 5110

### DISCUSSION THREAD

**Discussion by mikejhathaway** [OpenSC/OpenSC#2793] [General]
Hi Folks,

I have not been able to find a thread on this topic so please excuse me if this has come up before.

I am looking for a way to use OpenSC\P11 tool to initialize a Safenet 5110 USB smartcard, generate a private key & CSR that I can get signed by a CA, I have seen examples on Linux but I am looking for a solution on Windows, I am happy to go command line but I do not want to use CAPI.

So far I am able to initialize the token with the following, but am stuck with a solution for generating the private key etc

`** Initial the token, set SO PIN and PIN.

pkcs11-tool --module C:\Windows\System32\eTPKCS11.dll --init-token --label eToken_5110

pkcs11-tool --module C:\Windows\System32\eTPKCS11.dll --init-pin --login

** Generate a keypare with RSA2048.

pkcs11-tool --module C:\Windows\System32\eTPKCS11.dll --login --keypairgen --key-type RSA:2048 --id 1 --label "test@example.com"

** List the objects

pkcs11-tool --module C:\Windows\System32\eTPKCS11.dll --login --list-objects`

Any pointers would be appreciated

TIA

Mike

**Comment by dengert** [comment]
try Google for: `Safenet 5110 "pkcs11" "Windows" "eTPKCS11.dll"` 

What is output of the `pkcs11-tool --module C:\Windows\System32\eTPKCS11.dll --login --keypairgen --key-type RSA:2048 --id 1 --label "[test@example.com](mailto:test@example.com)"` and `pkcs11-tool --module C:\Windows\System32\eTPKCS11.dll --login --list-objects`

What version of OpenSC pkcs11-tool?

**Comment by mikejhathaway** [comment]
Hi @dengert,

Open SC is 0.23.0, I downloaded it yesterday, I am investigating using openssl engine_pkcs11, will have to compile for Win

Command output below

```
C:\Program Files\OpenSC Project\OpenSC\tools>pkcs11-tool --module C:\Windows\System32\eTPKCS11.dll --login --keypairgen --key-type RSA:2048 --id 1 --label "test@example.com"
Using slot 0 with a present token (0x0)
Logging in to "eToken_5110".
Please enter User PIN: Key pair generated:
Private Key Object; RSA
  label:      test@example.com
  ID:         01
  Usage:      decrypt, sign, unwrap
  Access:     sensitive, always sensitive, never extractable, local
Public Key Object; RSA 2048 bits
  label:      test@example.com
  ID:         01
  Usage:      encrypt, verify, wrap
  Access:     local
```

And

```
C:\Program Files\OpenSC Project\OpenSC\tools>pkcs11-tool --module C:\Windows\System32\eTPKCS11.dll --login --list-objects
Using slot 0 with a present token (0x0)
Logging in to "eToken_5110".
Please enter User PIN: Private Key Object; RSA
  label:      test@example.com
  ID:         01
  Usage:      decrypt, sign, unwrap
  Access:     sensitive, always sensitive, never extractable, local
Public Key Object; RSA 2048 bits
  label:      test@example.com
  ID:         01
  Usage:      encrypt, verify, wrap
  Access:     local
```

**Comment by dengert** [comment]
Not sure why you do not want to use CAPI, as 007-013560-006-SafeNet Authentication Client_10.8_R8_Windows_GA_Administrator_Guide_Rev B.pdf says eToken.dll is a wrapper for CAPI. And also says it can use CNG. 

engine should also work on Windows.

**Reply by mikejhathaway** [reply]
I don’t want to use CAPI because I have a P11 enabled application that needs to use keys on a token. CAPI and P11 use different key containers and formats, P11 token generated keys and certificates are not accessible when you take them to non-Windows platforms.

**Reply by dengert** [reply]
Really? I have not seen a smart card that is not usable by both. The token has the key and certificate on it.

Windows likes to import certificates into its certstore. So unlike pkcs11 it can find cert in the store and tell user which token to insert. I.e. uses windows containerid which is derived from token info and cert info on card. 

You can export cert from store which is der format.
